File: connectors/go/frame.go
package impulsering

import "encoding/binary"

const (
	frameMagic  = 0x5249 // "IR"
	wireVersion = 1
	frameHeader = 16
)

func encodeFrame(schemaFP uint64, body []byte) []byte {
	out := make([]byte, frameHeader+len(body))
	binary.LittleEndian.PutUint16(out[0:], frameMagic)
	out[2] = wireVersion
	out[3] = 0
	binary.LittleEndian.PutUint64(out[4:], schemaFP)
	binary.LittleEndian.PutUint32(out[12:], uint32(len(body)))
	copy(out[frameHeader:], body)
	return out
}

func decodeFrame(buf []byte) (schemaFP uint64, body []byte, err error) {
	if len(buf) < frameHeader {
		return 0, nil, errf("frame shorter than header")
	}
	if binary.LittleEndian.Uint16(buf[0:]) != frameMagic {
		return 0, nil, errf("bad frame magic")
	}
	if buf[2] != wireVersion {
		return 0, nil, errf("unsupported wire version %d", buf[2])
	}
	schemaFP = binary.LittleEndian.Uint64(buf[4:])
	bodyLen := int(binary.LittleEndian.Uint32(buf[12:]))
	if frameHeader+bodyLen > len(buf) {
		return 0, nil, errf("frame body truncated")
	}
	return schemaFP, buf[frameHeader : frameHeader+bodyLen], nil
}

File: connectors/python/impulse_ring/frame.py
"""Self-describing wire frame (see SPEC/wire-format.md section 3)."""

import struct

FRAME_MAGIC = 0x5249  # "IR"
WIRE_VERSION = 1
FRAME_HEADER = 16


def encode(schema_fp: int, body: bytes, flags: int = 0) -> bytes:
    return (
        struct.pack("<HBB", FRAME_MAGIC, WIRE_VERSION, flags)
        + struct.pack("<Q", schema_fp & ((1 << 64) - 1))
        + struct.pack("<I", len(body))
        + body
    )


def decode(buf: bytes):
    """Return (schema_fp, body) or raise ValueError."""
    if len(buf) < FRAME_HEADER:
        raise ValueError("frame shorter than header")
    magic, ver, _flags = struct.unpack_from("<HBB", buf, 0)
    if magic != FRAME_MAGIC:
        raise ValueError("bad frame magic")
    if ver != WIRE_VERSION:
        raise ValueError(f"unsupported wire version {ver}")
    (schema_fp,) = struct.unpack_from("<Q", buf, 4)
    (body_len,) = struct.unpack_from("<I", buf, 12)
    if FRAME_HEADER + body_len > len(buf):
        raise ValueError("frame body truncated")
    return schema_fp, bytes(buf[FRAME_HEADER : FRAME_HEADER + body_len])
